wrapper/src/main/java/software/amazon/jdbc/plugin/efm2/MonitorImpl.java

```diff
@@ -0,0 +1,418 @@
+
+  private static final Logger LOGGER = Logger.getLogger(MonitorImpl.class.getName());
+  private static final long THREAD_SLEEP_MILLIS = 1000;
+  private static final String MONITORING_PROPERTY_PREFIX = "monitoring-";
+
+  protected static final Executor ABORT_EXECUTOR = Executors.newSingleThreadExecutor();
+
+  private final Queue<WeakReference<MonitorConnectionContext>> activeContexts = new ConcurrentLinkedQueue<>();
+  private final HashMap<Long, Queue<WeakReference<MonitorConnectionContext>>> newContexts = new HashMap<>();
+  private final PluginService pluginService;
+  private final TelemetryFactory telemetryFactory;
+  private final Properties properties;
+  private final HostSpec hostSpec;
+  private final AtomicBoolean stopped = new AtomicBoolean(false);
+  private Connection monitoringConn = null;
+  private final ExecutorService threadPool = Executors.newFixedThreadPool(2, runnableTarget -> {
+    final Thread monitoringThread = new Thread(runnableTarget);
+    monitoringThread.setDaemon(true);
+    return monitoringThread;
+  });
+
+  private final long failureDetectionTimeNano;
+  private final long failureDetectionIntervalNano;
+  private final int failureDetectionCount;
+
+  private long invalidNodeStartTimeNano;
+  private long failureCount;
+  private boolean nodeUnhealthy = false;
+
+
+  private final TelemetryGauge newContextsSizeGauge;
+  private final TelemetryGauge activeContextsSizeGauge;
+  private final TelemetryGauge nodeHealtyGauge;
+  private final TelemetryCounter abortedConnectionsCounter;
+
+  /**
+   * Store the monitoring configuration for a connection.
+   *
+   * @param pluginService             A service for creating new connections.
+   * @param hostSpec                  The {@link HostSpec} of the server this {@link MonitorImpl}
+   *                                  instance is monitoring.
+   * @param properties                The {@link Properties} containing additional monitoring
+   *                                  configuration.
+   */
+  public MonitorImpl(
+      final @NonNull PluginService pluginService,
+      final @NonNull HostSpec hostSpec,
+      final @NonNull Properties properties,
+      final int failureDetectionTimeMillis,
+      final int failureDetectionIntervalMillis,
+      final int failureDetectionCount,
+      final TelemetryCounter abortedConnectionsCounter) {
+
+    this.pluginService = pluginService;
+    this.telemetryFactory = pluginService.getTelemetryFactory();
+    this.hostSpec = hostSpec;
+    this.properties = properties;
+    this.failureDetectionTimeNano = TimeUnit.MILLISECONDS.toNanos(failureDetectionTimeMillis);
+    this.failureDetectionIntervalNano = TimeUnit.MILLISECONDS.toNanos(failureDetectionIntervalMillis);
+    this.failureDetectionCount = failureDetectionCount;
+    this.abortedConnectionsCounter = abortedConnectionsCounter;
+
+    final String nodeId = StringUtils.isNullOrEmpty(this.hostSpec.getHostId())
+        ? this.hostSpec.getHost()
```

Comment:
Does it make sense to have move this check inside getHostId so we don't have to do this every call?
```java
  public String getHostId() {
    return StringUtils.isNullOrEmpty(hostId) ? host : hostId;
  }
```

Reply:
This code is in class constructor and runs just once. It we had many places with similar place then it'd make sense to extract such logic into a new method.